File: Utilities/bulk_insert.js
'use strict';

const path = require('path');
const fs = require('fs');
const commandLineArgs = require('command-line-args');
const prompt = require('prompt');
const mysql = require('mysql');

const dbCharset = 'utf8mb4';

// The defaultMaxAllowedPacket is the standard max_allowed_packet for our
// database. While performing the bulk insert operation, we increase
// the max_allowed_packet.
//
// When the database operations are complete we will restore the max_allowed_packet
// to the defaultMaxAllowedPacket.
//
// Default max_allowed_packet on our database is 64 megabytes
const defaultMaxAllowedPacket    =  67108864; // in bytes

// Prior to our bulk insert operation we set the max_allowed_packet to
// 500 MB
const bulkInsertMaxAllowedPacket = 524288000; // in bytes

const scriptName = path.basename(__filename);

// Each row consists of 1800 characters including the newline character
const fixedRowLength = 1800;

// maxBulkInsertCount is the maximum number of records to insert at a time.
const maxBulkInsertCount = 145554;

// Add one for the newline character
const maxFileLengthToRead = maxBulkInsertCount * ( fixedRowLength + 1 );

let password = null;
let dbConnection = null;
let filePosition = 0;

const optionDefinitions = [
    { name: 'input-file', alias: 'i', type: String },
    { name: 'db-host', alias: 'h', type: String, defaultOption: 'localhost' },
    { name: 'db-user', alias: 'u', type: String },
    { name: 'database', alias: 'd', type: String },
    { name: 'show-password', alias: 's', type: Boolean, defaultOption: false },
    { name: 'debug', alias: 'g', type: Boolean, defaultOption: false }
];

const commandLineOptions = commandLineArgs(optionDefinitions);
const promptPasswordSpec = {
    properties: {
        password: {
            pattern: /^[:print:\S]+$/,
            message: 'Password must consist of one or more non-whitespace ' +
                     'printable characters and cannot contain any whitespace ' +
                     'characters nor any non-printable characters.',
            required: true,
            hidden: commandLineOptions['show-password'] ? false : true
        }
    }
};

const debug = commandLineOptions['debug'] ? true : false;

function getDbConnection(host, user, database, password) {
    if (dbConnection) {
        return dbConnection;
    }
    if (!host) {
        throw new Error(
            `${scriptName} invoked without a valid --db-host`);
    } else if (!user) {
        throw new Error(
            `${scriptName} invoked without a valid --db-user`);
    } else if (!database) {
        throw new Error(
            `${scriptName} invoked without a valid --database`);
    } else if (!password) {
        throw new Error(
            `${scriptName} unable to connect ${user}@${host}:${database} : ` +
            `Password for database user is not specified.`)
    }

    dbConnection = mysql.createConnection({
        host: host,
        user: user,
        database: database,
        password: password,
        charset: dbCharset
    });

    if (!dbConnection) {
        throw new Error(
            `${scriptName} Failed to connect ${user}@${host}:${database}`);
    }

    return dbConnection;
}

function getDataObjectFromFixedLengthRecord(record, inputFileName, lineNumber) {
    const name = record.substr(0, 72).trim();
    if (!name) {
        console.warn(
            `Warning: ${scriptName}: Record on line number ${lineNumber} in ` +
            `file '${inputFileName}' is missing a valid consumer name.`);
    }
    const ssn = parseInt( record.substr(72, 9).trim().replace(/^0{1,8}/, ''), 10 );
    if (ssn.isNaN()) {
        throw new Error(
            `${scriptName}: Record on line number ${lineNumber} in file ` +
            `'${inputFileName}' is missing a valid social security number.`
        );
    }
    let = creditTagValues = [ null ];
    for (let creditTagNumber = 1; creditTagNumber <= 200; creditTagNumber++) {
        let startIndex = ( 72 + 9 ) + ( ( creditTagNumber - 1) * 9 );
        let value = record.substr( startIndex, 9 ).trim();
        if (value.substr( 0, 1) === '-') {
            value = value.replace(/^-0{1,7}/, '');
            if (value === '-0') {
                value = '0';
            }
        } else {
            value = value.replace(/^0{1,8}/, '');
        }
        value = parseInt( value, 10 );
        
    }
}

function extractRecordsToInsert(buffer, dataLength, inputFileName, startingLineCount) {
    let startBufferIndex = 0;
    let lineCount = startingLineCount - 1;

    while(startBufferIndex < dataLength) {
        lineCount += 1;
        let endBufferIndex = startBufferIndex + fixedRowLength + 1;
        let consumerRecord = buffer.toString('utf8', startBufferIndex, endBufferIndex);
    }
}

const getBulkInsertFunction = (options) => {
    const {
        connection,
        user,
        host,
        database,
        inputFile
    } = options;

    const bulkIsertFunction = (err) => {
        if (err) {
            console.error(`${scriptName}: Failed to connect ${user}@${host}:${database}: ${err}`);
            console.error(err.stack);
            throw err;
        }

        if (debug) {
            console.log(`${scriptName}: Connected ${user}@${host}:${database} as ID ${connection.threadId}`);
        }

        const readBuffer = Buffer.alloc(maxFileLengthToRead);

        let inputFd = null;
    
        try {
            inputFd = openSync(inputFile, 'r');
        } catch (err) {
            console.error(
                `Attempt to open the input file '${inputFile}' failed: ${err}`);
            throw err;
        }
    
        let endOfFileReached = false;
    
        while (!endOfFileReached) {
            let bytesRead = fs.readSync(
                                inputFd,
                                readBuffer,
                                0,
                                maxFileLengthToRead,
                                filePosition
                            );
            
            if (!bytesRead || bytesRead < maxFileLengthToRead) {
                endOfFileReached = true;
    
                if (!bytesRead) {
                    break;
                }
            }
        }
    

    }
    return bulkInsertFunction;
}

function main() {
    const inputFile = commandLineOptions['input-file'] ?
                      commandLineOptions['input-file'].trim() : null;

    if (!inputFile) {
        throw new Error(`${scriptName} invoked without a valid --input-file`);
    } else if (!fs.existsSync(inputFile)) {
        throw new Error(
            `${scriptName} invoked with --input-file='${inputFile}' : ` +
            `Specified Input file does not exist.`);
    }

    const inputFileStats = fs.statSync(inputFile);

    if (inputFileStats.isDirectory()) {
        throw new Error(
            `${scriptName} invoked with --input-file='${inputFile}' : ` +
            `Specified Input file is a directory.`);
    }

    const connection = getDbConnection(commandLineOptions['db-host'],
                                       commandLineOptions['db-user'],
                                       commandLineOptions['database'],
                                       password);
    
    const readBuffer = Buffer.alloc(maxFileLengthToRead);

    let inputFd = null;

    try {
        inputFd = openSync(inputFile, 'r');
    } catch (err) {
        console.error(
            `Attempt to open the input file '${inputFile}' failed: ${err}`);
        throw err;
    }

    let endOfFileReached = false;

    while (!endOfFileReached) {
        let bytesRead = fs.readSync(
                            inputFd,
                            readBuffer,
                            0,
                            maxFileLengthToRead,
                            filePosition
                        );
        
        if (!bytesRead || bytesRead < maxFileLengthToRead) {
            endOfFileReached = true;

            if (!bytesRead) {
                break;
            }
        }
    }

}

// Password should never be specified as a command-line command option.
// Password may be in a file and the file may be specified on the
// command-line command.
// User may be prompted for their password.
// The reason we never retrieve the password from a command-line command option
// is that the command gets logged in a variety of places, including in history.
// If we permit the password to be specified as a command-line command option in
// clear text, then the password will be logged in a variety of places including
// history.
prompt.get(promptPasswordSpec,
           (err, result) => {
               if (err) {
                   throw new Error(
                       `${scriptName} error occurred while prompting ` +
                       `command-line user for their database password: ${err}`);
               }
               password = result.password;
               main();
           });

File: Utilities/bulk_insert.py
import sys
import os
import os.path
import re
import argparse
import pgdb

SCRIPT_NAME = re.sub(r'^.*\/', '', os.path.abspath(sys.argv[0]));

arg_parser = argparse.ArgumentParser(
exit_on_error=True,
description="""\
'This Python program takes in an input file that contains a series of records,
one for each consumer. Each line contains the consumer name, social security
number, credit tags X0001 through X0200.

Each line consists of 1881 characters (consumer name - 72 characters,
social security number - 9 characters, and credit tag metrics for 200 credit
tags, each of 9 characters in length) followed by 1 if newline character.

We read in blocks of data from the input file and then perform bulk inserts.
"""
)

# Add argument: --input-file
arg_parser.add_argument(
    '--input-file',
    metavar='INPUT-FILE',
    type=str,
    required=True,
    help='Required: specifies the input file to load the data from for ' \
         'executing the bulk insert operation.'
)

# Add argument: --db-host
arg_parser.add_argument(
    '--db-host',
    metavar='HOST',
    type=str,
    default='localhost',
    help='Specifies the host that is hosting our database. If not specified,' \
         'then localhost is used.'
)

# Add argument: --db-port
arg_parser.add_argument(
    '--db-port',
    metavar='PORT',
    type=str,
    help='Specifies the port that is used for connecting to our database. ' \
         'If not specified, then the PostGreSQL default port is used for ' \
         'the connection.'
)

# Add argument: --db-user
arg_parser.add_argument(
    '--db-user',
    metavar='USER',
    type=str,
    required=True,
    help='Required: Specifies the user for our database.'
)

# Add argument: --database
arg_parser.add_argument(
    '--database',
    metavar='DATABASE',
    type=str,
    required=True,
    help='Required: Specifies the name of our database.'
)

# Add argument: --debug
arg_parser.add_argument(
    '--debug',
    help=f"When --debug is used {SCRIPT_NAME} will output debug information " \
           "to STDERR. When --debug is not used, then debug information will " \
           "be suppressed."
)

PROG_ARGS_NAMESPACE = arg_parser.parse_args()
PROG_ARGS = vars(PROG_ARGS_NAMESPACE)

# The defaultMaxAllowedPacket is the standard max_allowed_packet for our
# database. While performing the bulk insert operation, we increase
# the max_allowed_packet.
#
# When the database operations are complete we will restore the max_allowed_packet
# to the defaultMaxAllowedPacket.
#
# Default DEFAULT_MAX_ALLOWED_PACKET on our database is 64 megabytes
DEFAULT_MAX_ALLOWED_PACKET    =   67108864  # in bytes

# Prior to our bulk insert operation we set the max_allowed_packet to
# 500 MB
BULK_INSERT_MAX_ALLOWED_PACKET = 524288000  # in bytes

# Each row consists of 1881 characters including the newline character
FIXED_ROW_LENGTH = 1881;

# maxBulkInsertCount is the maximum number of records to insert at a time.
MAX_BULK_INSERT_COUNT = 145554;

# Add one for the newline character
MAX_FILE_LENGTH_TO_READ = MAX_BULK_INSERT_COUNT * ( FIXED_ROW_LENGTH + 1 );

def main():
    input_file = PROG_ARGS.get('input_file')
    if not input_file:
        raise Exception(f"{SCRIPT_NAME} invoked without a valid --input-file.")
    elif not os.path.exists(input_file):
        raise Exception(f"{SCRIPT_NAME} invoked with " \
                        f"--input-file='{input_file}'. However, {input_file} " \
                        "does not exist or cannot be read from. Cannot continue")
    elif os.path.isdir(input_file):
        raise Exception(f"{SCRIPT_NAME} invoked with " +
                        f"--input-file='{input_file}'. However, {input_file} " \
                        "exists but is a directory and not a file. " \
                        "Cannot continue")


main()
